File: app/src/main/java/com/brickgit/imagesearch/ui/VideoFragment.java
package com.brickgit.imagesearch.ui;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.brickgit.imagesearch.R;
import com.brickgit.imagesearch.adapter.SpaceItemDecoration;
import com.brickgit.imagesearch.adapter.VideoAdapter;
import com.brickgit.imagesearch.model.MediaViewModel;

import butterknife.BindView;
import butterknife.ButterKnife;

public class VideoFragment extends Fragment {


	private MediaViewModel mViewModel;

	@BindView(R.id.video_progress_bar) ProgressBar progressBar;
	@BindView(R.id.video_list_view) RecyclerView listView;

	private StaggeredGridLayoutManager layoutManager;

	private final VideoAdapter.OnVideoCellClickListener onVideoCellClickListener = videoInfo -> {
		Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(videoInfo.pageURL));
		startActivity(intent);
	};

	private final RecyclerView.OnScrollListener onScrollListener = new RecyclerView.OnScrollListener() {
		@Override
		public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
			super.onScrolled(recyclerView, dx, dy);

			int totalItemCount = layoutManager.getItemCount();
			int visibleItemCount = layoutManager.getChildCount();
			int[] firstVisibleItemPositions = layoutManager.findFirstVisibleItemPositions(null);
			int firstVisibleItemPosition = firstVisibleItemPositions[firstVisibleItemPositions.length - 1];

			if (firstVisibleItemPosition + visibleItemCount >= totalItemCount) {
				mViewModel.loadMoreVideos();
			}
		}
	};

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_video, container, false);
		ButterKnife.bind(this, view);

		layoutManager = new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL);
		listView.setLayoutManager(layoutManager);
		VideoAdapter adapter = new VideoAdapter();
		adapter.setOnVideoCellClickListener(onVideoCellClickListener);
		listView.setAdapter(adapter);
		listView.addItemDecoration(new SpaceItemDecoration());
		listView.addOnScrollListener(onScrollListener);

		mViewModel = ViewModelProviders.of(getActivity()).get(MediaViewModel.class);
		mViewModel.getQuery().observe(this, query -> {
			showProgressBar(true);
			adapter.clear();
		});
		mViewModel.getVideos().observe(this, imageInfoResponses -> {
			showProgressBar(false);
			adapter.update(imageInfoResponses);
		});

		return view;
	}

	private void showProgressBar(boolean show) {
		progressBar.setVisibility(show ? View.VISIBLE : View.INVISIBLE);
	}
}

File: app/src/main/java/com/guanhaolin/pearch/ui/video/VideoFragment.kt
package com.guanhaolin.pearch.ui.video

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.guanhaolin.pearch.api.model.VideoResponse
import com.guanhaolin.pearch.databinding.FragmentVideoBinding
import com.guanhaolin.pearch.ui.MediaViewModel
import com.guanhaolin.pearch.ui.util.SpaceItemDecoration
import com.guanhaolin.pearch.ui.video.VideoAdapter.OnVideoCellClickListener
import org.koin.androidx.viewmodel.ext.android.activityViewModel

class VideoFragment : Fragment() {

    private var _binding: FragmentVideoBinding? = null
    private val binding get() = _binding!!

    private val viewModel: MediaViewModel by activityViewModel()

    private lateinit var layoutManager: StaggeredGridLayoutManager
    private lateinit var adapter: VideoAdapter

    private val onVideoCellClickListener =
        object : OnVideoCellClickListener {
            override fun onVideoCellClicked(video: VideoResponse) {
                val intent = Intent(Intent.ACTION_VIEW, Uri.parse(video.pageURL))
                startActivity(intent)
            }
        }

    private val onScrollListener: RecyclerView.OnScrollListener =
        object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                val totalItemCount = layoutManager.itemCount
                val visibleItemCount = layoutManager.childCount
                val firstVisibleItemPositions = layoutManager.findFirstVisibleItemPositions(null)
                val firstVisibleItemPosition =
                    firstVisibleItemPositions[firstVisibleItemPositions.size - 1]

                if (firstVisibleItemPosition + visibleItemCount >= totalItemCount) {
                    viewModel.loadMoreVideos();
                }
            }
        }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        _binding = FragmentVideoBinding.inflate(inflater, container, false).apply {
            layoutManager = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL)
            videoListView.setLayoutManager(layoutManager)
            adapter = VideoAdapter()
            adapter.setOnVideoCellClickListener(onVideoCellClickListener)
            videoListView.setAdapter(adapter)
            videoListView.addItemDecoration(SpaceItemDecoration())
            videoListView.addOnScrollListener(onScrollListener)
        }

        viewModel
            .uiState
            .observe(
                viewLifecycleOwner
            ) { uiState ->
                val videos = uiState.videos
                showProgressBar(videos.isLoading && videos.data.isEmpty())
                adapter.update(videos.data)
            }

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun showProgressBar(show: Boolean) {
        binding.videoProgressBar.visibility = if (show) View.VISIBLE else View.GONE
    }
}
